Add `force_regular` property and fine-tune light vs regular behavior

ljfranklin opened the pull request:
- If `force_regular` is true (default false), the resource will only check/get
  stemcells with that have a regular (i.e. heavy) stemcell URL
- Handles timing issues in stemcell publishing process where a heavy stemcell appears on
  bosh.io before the corresponding light stemcell
  - Unless `force_regular` is true, `check` will now ignore new versions that only have a regular
    stemcell and will only trigger once the light version appears for IaaSes that support light stemcells
  - It would be bad on AWS if Concourse triggered on a heavy stemcell instead of waiting for the light
    stemcell

[#129544375](https://www.pivotaltracker.com/story/show/129544375)

ljfranklin committed:
Add `force_regular` property and fine-tune light vs regular behavior

- If `force_regular` is true (default false), the resource will only check/get
  stemcells with that have a regular (i.e. heavy) stemcell URL
- Handles timing issues in stemcell publishing process where a heavy stemcell appears on
  bosh.io before the corresponding light stemcell
  - Unless `force_regular` is true, `check` will now ignore new versions that only have a regular
    stemcell and will only trigger once the light version appears for IaaSes that support light stemcells
  - It would be bad on AWS if Concourse triggered on a heavy stemcell instead of waiting for the light
    stemcell

[#129544375](https://www.pivotaltracker.com/story/show/129544375)

joshzarrabi reviewed (commented):
This seems well tested. Could be useful for some users.

ljfranklin commented:
@vito Think we could get this into the next Concourse release? We're about to get light GCE stemcells up on bosh.io. The light versions are published after the heavy version, so everyone's pipeline will pull the heavy version until this PR is merged. Thanks!

vito commented:
@ljfranklin Probably, wasn't sure if I should review or @zachgersh since this resource is kinda his baby now. :) I'll take a look myself at least but wait for him before merging.

vito commented on cmd/check/main.go:
```diff
@@ -32,12 +33,27 @@ func main() {
+			stemcells = client.FilterStemcells(regularFilter, stemcells)
+		} else {
+			lightFilter := func(stemcell boshio.Stemcell) bool {
+				return stemcell.Light != nil
```
Hmm, should this instead be a no-op (`return true`)? It seems like it'd exclusively only match stemcells with a light version available. What if it's an IaaS that only ever has regular stemcells and `force_regular` is false?

ljfranklin replied:
@vito the `supportsLight` returns true if that IaaS supports light stemcells, so we don't execute the `ForceRegular` check for IaaS that only have regular stemcells.

Cases:
- IaaS only supports regular, don't both filtering
- IaaS supports light and ForceRegular is true, only return Regular stemcells
  - This is used in our light stemcell builder pipeline to only fetch heavy stemcells
- IaaS supports light and ForceRegular is false, only return Light stemcells
  - This ensures the resource waits until the light version is out before running. This is important as the heavy GCE stemcells will be available an hour or so before the light versions.

vito replied:
Ah, gotcha.

zachgersh replied:
I'd actually like to request that we change this. Instead of the complexity of having to call what is a filtering function a `lamba` can't we just move all of the filtering logic into the function that is actually there to filter stemcells? To me, it reduces the complexity of the main.go and pushes the responsibility down to the package that should be holding onto it.

zachgersh commented:
@ljfranklin / @vito my review was `coming soon`. I will make sure to throw it up tonight 💃

zachgersh commented on acceptance/check_test.go:
```diff
@@ -64,10 +89,72 @@ var _ = Describe("check", func() {
+	})
+
+	Context("when `force_regular` is true", func() {
+
```
please kill this space

zachgersh commented on acceptance/check_test.go:
```diff
@@ -64,10 +89,72 @@ var _ = Describe("check", func() {
+
+				result := []stemcellVersion{}
+				err = json.Unmarshal(session.Out.Contents(), &result)
+				Expect(err).ToNot(HaveOccurred())
```
can you please change these to `NotTo` just for consistency across the whole project. Don't want to flip back and forth.

zachgersh commented on cmd/in/main.go:
```diff
@@ -55,17 +56,21 @@ func main() {
-		log.Fatalln(err)
+	filterFunc := func(stemcell boshio.Stemcell) bool {
+		return stemcell.Version == inRequest.Version.Version
+	}
```
Again, I'd love to figure out how to push this logic down into the `FIlterStemcell` function itself instead of it living in main.

zachgersh commented on versions/versions.go:
```diff
@@ -22,6 +22,10 @@ func NewFilter(initialVersion string, stemcells []boshio.Stemcell) Filter {
+	if len(f.stemcells) == 0 {
+		return []List{}, nil
+	}
+
```
I like this but when do you actually forsee this case happening?

cunnie replied:
@zachgersh In practice we shouldn't hit this case, just being defensive.

zachgersh reviewed (changes requested):
Hey lyle,

Really appreciate these changes and apologies that it took me so long to review :/  I a few small changes but I think this is great.

zachgersh commented:
@ljfranklin what do you think of also making the error more explicit if you have `force_regular` false but then are trying to get something like the azure stemcell (which only has a heavy). It could see someone potentially not understanding that they need to make sure to set this option to true.

Commit:
Incorporate PR feedback from Zach and Alex

- ToNot -> NotTo
- Remove extra spaces
- Pushed filtering logic into new `Stemcells` struct

Signed-off-by: Lyle Franklin <lfranklin@pivotal.io>

[#129544375](https://www.pivotaltracker.com/story/show/129544375)

cunnie commented:
@zachgersh @vito thanks for the feedback. We pushed our updates, please give it another glance to make sure we didn't miss anything.

@zachgersh To your point about throwing an error if `force_regular` is set `false` and there are no light stemcells (e.g. Azure): the user will get the regular (non-light) stemcell. The intent of `force_regular` is to override the default behavior of pulling light stemcells if the IaaS supports light stemcells, otherwise grab the regular stemcells. For Azure, the code will see that there are no light stemcells and happily grab the heavy stemcells, even if `force_regular` is `false`. The main use case for `force_regular` is something like our light stemcell builder pipeline which needs the regular stemcell even though the IaaS supports light stemcells.

zachgersh commented:
@cunnie thanks to you and @ljfranklin I am going to go ahead and pull this in. Any other changes you want to see let me know!
